pkg/plugin/keyvault.go

```diff
@@ -94,37 +115,54 @@ func newKeyVaultClient(
 
 	klog.InfoS("using kms key for encrypt/decrypt", "vaultURL", *vaultURL, "keyName", keyName, "keyVersion", keyVersion)
 
-	client := &keyVaultClient{
+	client := &KeyVaultClient{
 		baseClient:       kvClient,
 		config:           config,
 		vaultName:        vaultName,
 		keyName:          keyName,
 		keyVersion:       keyVersion,
 		vaultURL:         *vaultURL,
 		azureEnvironment: env,
+		keyID:            fmt.Sprintf("%x", sha256.Sum256([]byte(fmt.Sprintf("%skeys/%s/%s", *vaultURL, keyName, keyVersion)))),
+		algorithm:        kv.RSA15,
 	}
 	return client, nil
 }
 
-func (kvc *keyVaultClient) Encrypt(ctx context.Context, cipher []byte) ([]byte, error) {
-	value := base64.RawURLEncoding.EncodeToString(cipher)
+// Encrypt encrypts the given plain text using the keyvault key.
+func (kvc *KeyVaultClient) Encrypt(ctx context.Context, plain []byte) (*service.EncryptResponse, error) {
+	value := base64.RawURLEncoding.EncodeToString(plain)
 
 	params := kv.KeyOperationsParameters{
-		Algorithm: kv.RSA15,
+		Algorithm: kvc.algorithm,
 		Value:     &value,
 	}
 	result, err := kvc.baseClient.Encrypt(ctx, kvc.vaultURL, kvc.keyName, kvc.keyVersion, params)
 	if err != nil {
 		return nil, fmt.Errorf("failed to encrypt, error: %+v", err)
 	}
-	return []byte(*result.Result), nil
+
+	annotations := map[string][]byte{
+		dateAnnotationKey:           []byte(result.Header.Get(dateAnnotationValue)),
+		requestIDAnnotationKey:      []byte(result.Header.Get(requestIDAnnotationValue)),
+		keyvaultRegionAnnotationKey: []byte(result.Header.Get(keyvaultRegionAnnotationValue)),
+		versionAnnotationKey:        []byte(encryptionResponseVersion),
+		algorithmAnnotationKey:      []byte(kvc.algorithm),
+	}
+
+	return &service.EncryptResponse{
+		Ciphertext:  []byte(*result.Result),
+		KeyID:       kvc.keyID,
+		Annotations: annotations,
+	}, nil
 }
 
-func (kvc *keyVaultClient) Decrypt(ctx context.Context, plain []byte) ([]byte, error) {
-	value := string(plain)
+// Decrypt decrypts the given cipher text using the keyvault key.
+func (kvc *KeyVaultClient) Decrypt(ctx context.Context, cipher []byte) ([]byte, error) {
```

Comment:
I would remove `ValidateAnnotations` from the interface and call it directly here so that it is impossible to forget to do.